datacube/index/abstract/_metadata_types.py

```diff
@@ -50,7 +50,7 @@ def from_doc(self, definition: JsonDict) -> MetadataType:
     def add(self,
             metadata_type: MetadataType,
             allow_table_lock: bool = False
-           ) -> MetadataType:
+           ) -> MetadataType | None:
```

Comment:
The subclasses were all returning None, so update the abstract methods for these instead.

Reply:
The subclasses end with `return get_by_name(...)` which does indeed return `MetadataType | None` - but `None` means the type doesn't exist in the database - which could only happen if there was a database error on insert, which would have raised an exception.

So we should probably be casting those return statements in the subclasses to `MetadataType` rather than declaring  `MetadataType | None` here.

Other than that, I'm ready to approve and merge this.

Reply:
I'd like to try a less heavy-handed approach, but that is beyond the scope of this PR. To make this 1000+ line change mergeable, I removed the return types I added on these implementations and aligned the types in the `_metadata_types.py` implementations with the types in the abstract class.

Since I rebased this and added some trivial types in some test parts, here's the relevant diff compared to what you had just reviewed when you wrote your comment:
https://github.com/opendatacube/datacube-core/compare/ce1f7a685b404329c6d60c199478f1e1847fbd26..9be620554c55fecd0ef3d6dcf03c167e5cffe107
The only difference besides the added types in `_metadata_types.py` is fixing the return type of `_make_many` to correspond with the stated return type in the PyDoc.

Reply:
Thanks, that's really helpful.